File: apps/fastapi/app/models/activity_gallery.py
from datetime import date, datetime

from sqlalchemy import BigInteger, Boolean, Column, Date, DateTime, ForeignKey, Integer, LargeBinary, String, Text
from sqlalchemy.orm import relationship

from app.core.database import Base


class ActivityGallery(Base):
    __tablename__ = "activity_gallery"

    id = Column(BigInteger, primary_key=True, autoincrement=True, index=True)
    tenant_id = Column(BigInteger, ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    gallery_name = Column(String(255), nullable=False)
    gallery_type = Column(String(20), nullable=False)
    description = Column(Text, nullable=True)
    activity_date = Column(Date, nullable=False)
    created_by = Column(BigInteger, nullable=False)
    is_published = Column(Boolean, default=False, nullable=False)
    status = Column(Integer, default=1, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    media_items = relationship(
        "ActivityGalleryMedia",
        back_populates="gallery",
        cascade="all, delete-orphan",
        lazy="select",
    )
    class_mappings = relationship(
        "ActivityGalleryClassMapping",
        back_populates="gallery",
        cascade="all, delete-orphan",
        lazy="select",
    )


class ActivityGalleryMedia(Base):
    __tablename__ = "activity_gallery_media"

    id = Column(BigInteger, primary_key=True, autoincrement=True, index=True)
    gallery_id = Column(
        BigInteger,
        ForeignKey("activity_gallery.id", ondelete="CASCADE"),
        nullable=False,
    )
    media_type = Column(String(20), nullable=False)
    file_name = Column(String(255), nullable=False)
    original_file_name = Column(String(255), nullable=True)
    file_path = Column(String(1000), nullable=False)
    file_content = Column(LargeBinary, nullable=True)
    file_size = Column(BigInteger, nullable=True)
    display_order = Column(Integer, default=1, nullable=False)
    uploaded_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    status = Column(Integer, default=1, nullable=False)

    gallery = relationship("ActivityGallery", back_populates="media_items")


class ActivityGalleryClassMapping(Base):
    __tablename__ = "activity_gallery_class_mapping"

    id = Column(BigInteger, primary_key=True, autoincrement=True, index=True)
    gallery_id = Column(
        BigInteger,
        ForeignKey("activity_gallery.id", ondelete="CASCADE"),
        nullable=False,
    )
    class_id = Column(BigInteger, nullable=False)
    division_id = Column(BigInteger, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    gallery = relationship("ActivityGallery", back_populates="class_mappings")

File: apps/web/src/types/activityGallery.ts
export type GalleryType = "Photo" | "Video";
export type MediaType = "Photo" | "Video";

export interface ActivityGalleryMedia {
  id: number;
  gallery_id: number;
  media_type: MediaType;
  file_name: string;
  original_file_name?: string | null;
  file_path: string;
  file_size?: number | null;
  display_order: number;
  uploaded_at: string;
}

export interface ActivityGalleryClassMapping {
  id: number;
  gallery_id: number;
  class_id: number;
  division_id: number;
  class_name?: string | null;
  division_name?: string | null;
  created_at?: string | null;
}

export interface ActivityGallery {
  id: number;
  tenant_id: number;
  gallery_name: string;
  gallery_type: GalleryType;
  description?: string | null;
  activity_date: string;
  created_by: number;
  is_published: boolean;
  status: number;
  created_at: string;
  updated_at: string;
  class_id?: number | null;
  class_name?: string | null;
  division_id?: number | null;
  division_name?: string | null;
  photo_count: number;
  video_count: number;
  media_count: number;
  class_mappings?: ActivityGalleryClassMapping[];
  media_items?: ActivityGalleryMedia[];
}

export interface ActivityGalleryListItem {
  id: number;
  tenant_id: number;
  gallery_name: string;
  gallery_type: GalleryType;
  activity_date: string;
  description?: string | null;
  is_published: boolean;
  created_at: string;
  updated_at: string;
  class_id?: number | null;
  class_name?: string | null;
  division_id?: number | null;
  division_name?: string | null;
  photo_count: number;
  video_count: number;
  media_count: number;
}

export interface ActivityGalleryListResponse {
  data: ActivityGalleryListItem[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface ActivityGalleryClassTarget {
  class_id: number;
  division_id: number;
}

export interface ActivityGalleryCreateRequest {
  gallery_name: string;
  gallery_type: GalleryType;
  activity_date: string;
  description?: string | null;
  targets: ActivityGalleryClassTarget[];
}

export interface ActivityGalleryUpdateRequest {
  gallery_name?: string;
  activity_date?: string;
  description?: string | null;
  targets?: ActivityGalleryClassTarget[];
}

export interface ActivityGalleryPublishResponse {
  message: string;
  gallery: ActivityGallery;
}

export interface ActivityGalleryDeleteResponse {
  message: string;
}

export interface ClassOption {
  id: number;
  name: string;
}

export interface DivisionOption {
  id: number;
  division_name: string;
}

export interface TeacherGalleryClassDivision {
  id: number;
  division_name: string;
}

export interface TeacherGalleryClassOption {
  id: number;
  name: string;
  divisions: TeacherGalleryClassDivision[];
}

export interface TeacherGalleryScopeResponse {
  is_teacher: boolean;
  default_targets: ActivityGalleryClassTarget[];
  classes: TeacherGalleryClassOption[];
}

export interface GalleryAccessPermissions {
  can_view: boolean;
  can_create: boolean;
  can_edit: boolean;
  can_delete: boolean;
  can_download: boolean;
}
